### PR TITLE
#3432 gw command topics

### PR DESCRIPTION
**Pull request by n-deliyski**

### CODE REVIEW

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java [hunk 1]
```diff
@@ -1546,6 +1546,16 @@ private Future<ProtocolAdapterCommandConsumer> createCommandConsumer(final Comma
...
                                 commandContext.getTracingContext());
+                    } else {
+                        // in case of gateway having subscribed for itself as simple device do not forward messages for
+                        // devices behind that gateway
+                        if (!subscription.isGatewaySubscriptionForAllDevices() && command.isTargetedAtGateway()) {
+                            log.debug(
+                                    "no command subscription via gateway found [tenant: {}, device-id: {}, gateway-id: {}]",
+                                    command.getTenant(), command.getDeviceId(), command.getGatewayId());
+                            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                    "no command subscription via gateway"));
```
Comment (calohmn) [inline adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java:1557]: I think the exception message here and the log message above could be improved.
Something like "Mapped target gateway not subscribed to commands for devices behind that gateway", or "Mapped target gateway only subscribed to commands for itself".

**Comment by calohmn** [comment]
Linking fixed issue: #3432.

File: site/documentation/content/user-guide/mqtt-adapter.md [hunk 2]
```diff
@@ -608,6 +608,9 @@ that gateway has created such a command subscription, that gateway will be chose
 on behalf of the command target device and that has an open subscription will be chosen randomly to receive the command
 message.
 
+It is not possible to use both topic names `command//+/req/#` and `command//[<gateway-id>]/req/#` at the same time. As
+at that case only the last such subscription would be handled from the Command Router.
+
```
Comment (calohmn) [inline site/documentation/content/user-guide/mqtt-adapter.md:613]: I would suggest replacing `As at that case` with `In that case` here.

File: site/homepage/content/release-notes.md [hunk 3]
```diff
@@ -4,6 +4,14 @@ title = "What is new & noteworthy in Hono?"
 description = "Information about changes in recent Hono releases. Includes new features, fixes, enhancements and API changes."
 +++
 
+## 2.3.0
+
+### Fixes & Enhancements
+
+* The MQTT adapter did forward command messages for devices behind gateway when the gateway subscription was targeting
```
Comment (calohmn) [inline site/homepage/content/release-notes.md:11]: "behind _a_ gateway"

File: site/homepage/content/release-notes.md [hunk 4]
```diff
@@ -4,6 +4,14 @@ title = "What is new & noteworthy in Hono?"
...
+the gateway itself with topic name `command//[<gateway-id>]/req/#`. This has been fixed and the commands for devices,
+for which the gateway acts on its behalf of, are not forwarded to the gateway.
```
Comment (calohmn) [inline site/homepage/content/release-notes.md:13]: The last sentence sounds a bit complicated. I would suggest:
"This has been fixed. For a subscription with such a topic name, only commands for the gateway itself are forwarded, not commands for devices, that the gateway acts on behalf of.

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java [hunk 5]
```diff
@@ -1546,6 +1546,16 @@ private Future<ProtocolAdapterCommandConsumer> createCommandConsumer(final Comma
...
                                 authenticatedDevice,
                                 commandContext.getTracingContext());
+                    } else {
+                        // in case of gateway having subscribed for itself as simple device do not forward messages for
+                        // devices behind that gateway
+                        if (!subscription.isGatewaySubscriptionForAllDevices() && command.isTargetedAtGateway()) {
+                            log.debug(
+                                    "no command subscription via gateway found [tenant: {}, device-id: {}, gateway-id: {}]",
+                                    command.getTenant(), command.getDeviceId(), command.getGatewayId());
+                            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
```
[the remark this answers is not in the document]
Comment (n-deliyski) [reply adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java:1556]: Agree, I was wondering what type of error to provide and `NoConsumerException` suites better.

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java (hunk 1 above)
Comment (n-deliyski) [reply adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java]: done

File: site/documentation/content/user-guide/mqtt-adapter.md (hunk 2 above)
Comment (n-deliyski) [reply site/documentation/content/user-guide/mqtt-adapter.md]: done

File: site/homepage/content/release-notes.md (hunk 4 above)
Comment (n-deliyski) [reply site/homepage/content/release-notes.md]: Right, your suggestion is better.

**Comment by calohmn** [comment]
@n-deliyski Can you rebase?

**Comment by n-deliyski** [comment]
@calohmn The rebasing is done.

**Review by calohmn — APPROVED** [review]
LGTM

**Pull request merged**